src/passes/OptimizeInstructions.cpp

```diff
@@ -1224,14 +1230,78 @@ struct OptimizeInstructions
             !EffectAnalyzer(getPassOptions(), features, left->right)
                .hasSideEffects()) {
           switch (left->op) {
-            //   (x > y) | (x == y)    ==>    x >= y
             case EqInt32: {
-              if (right->op == GtSInt32) {
-                left->op = GeSInt32;
-                return left;
+              switch (right->op) {
+                //   (x >  y) | (x == y)   ==>    x >= y
+                //   (x >= y) | (x == y)   ==>    x >= y
+                case GtSInt32:
+                case GeSInt32: {
+                  left->op = GeSInt32;
+                  return left;
+                }
+                case GtUInt32:
+                case GeUInt32: {
+                  left->op = GeUInt32;
+                  return left;
+                }
+                //   (x  < y) | (x == y)    ==>    x <= y
+                //   (x <= y) | (x == y)    ==>    x <= y
+                case LtSInt32:
+                case LeSInt32: {
+                  left->op = LeSInt32;
+                  return left;
+                }
+                case LtUInt32:
+                case LeUInt32: {
+                  left->op = LeUInt32;
+                  return left;
+                }
+                //   (x != y) | (x == y)   ==>    1
+                case NeInt32: {
+                  return LiteralUtils::makeFromInt32(
+                    1, Type::i32, *getModule());
+                }
+                default: {
+                }
+              }
+              break;
+            }
+            case LtSInt32:
+            case LtUInt32: {
+              switch (right->op) {
+                //   (x > y) | (x < y)    ==>    x != y
+                case GtSInt32:
+                case GtUInt32: {
```

Comment:
don't these need to match in sign? `x (signed)> y || x (unsigned)< y` isn't the same as `x != y`, for example `x = -1, y = 1` (false before this opt, true after)

Reply:
Good point! Will add necessary checks